File: app/src/main/java/com/ed/redditapp/ui/search/SearchAsyncTask.java
package com.ed.redditapp.ui.search;

import com.ed.redditapp.lib.AsyncTask;
import com.ed.redditapp.lib.Timer;
import com.ed.redditapp.lib.api.RedditApi;
import com.ed.redditapp.lib.api.SearchItemSubreddit;

public class SearchAsyncTask {
    public interface Callback {
        void callback(SearchItemSubreddit[] result);
    }

    private final Timer searchDelayTimer;

    private String searchText;
    private Callback searchCallback;

    public SearchAsyncTask(RedditApi api) {
        searchDelayTimer = new Timer(1000, () -> {
            new AsyncTask().runAsync(h -> {
                SearchItemSubreddit[] result = api.searchSubreddits(searchText);
                h.onUIThread(() -> searchCallback.callback(result));
            });
        });
    }

    public void run(String searchText, Callback searchCallback) {
        this.searchText = searchText;
        this.searchCallback = searchCallback;
        searchDelayTimer.run();
    }
}

File: app/src/main/java/com/ed/redditapp/ui/search/SearchAsyncTask.kt
package com.ed.redditapp.ui.search

import com.ed.redditapp.lib.AsyncTask
import com.ed.redditapp.lib.AsyncTask.UIHandler
import com.ed.redditapp.lib.Timer
import com.ed.redditapp.lib.api.RedditApi
import com.ed.redditapp.lib.api.SearchItemSubreddit

typealias SearchAsyncCallback = (result: Array<SearchItemSubreddit>) -> Unit

class SearchAsyncTask(api: RedditApi) {
    private val searchDelayTimer = Timer(1000) {
        AsyncTask().runAsync { h: UIHandler ->
            val result = api.searchSubreddits(searchText!!)
            h.onUIThread { searchCallback!!(result) }
        }
    }

    private var searchText: String? = null
    private var searchCallback: SearchAsyncCallback? = null

    fun run(searchText: String, callback: SearchAsyncCallback?) {
        this.searchText = searchText
        this.searchCallback = callback
        searchDelayTimer.run()
    }
}